index.bs

```diff
@@ -2202,6 +2202,45 @@ During the above process, the user agent SHOULD show some UI to the user to guid
 authorizing an authenticator.
 </div>
 
+#### Registration API Exceptions
+
+[=[WRPS]=] can encounter the following {{DOMException|DOMExceptions}} from a call to {{CredentialsContainer/create()|navigator.credentials.create()}}. Some errors can have multiple reasons for why they happened, requiring the [=[WRPS]=] to infer the actual reason based on their use of WebAuthn:
+
+<dl dfn-type="argument" dfn-for="Registration API Exceptions">
+    :   <dfn>AbortError</dfn>
+    ::  The ceremony was cancelled via an {{AbortController}}
+        (see [[#sctn-abortoperation]] and [[#sctn-sample-aborting]] for more information.)
+
+    :   <dfn>ConstraintError</dfn>
+    ::  Either {{residentKey}} was set to "{{ResidentKeyRequirement/required}}" and no available authenticator supported resident keys,
+        or {{userVerification}} was set to "{{UserVerificationRequirement/required}}" and no available authenticator could perform [=user verification=].
+
+    :   <dfn>InvalidStateError</dfn>
+    ::  The authenticator used in the ceremony recognized an entry in {{excludeCredentials}}
+        after the user [=user consent|consented=] to registering a credential.
+
+    :   <dfn>NotSupportedError</dfn>
+    ::  No entry in {{PublicKeyCredentialCreationOptions/pubKeyCredParams}} had a `type` property of "{{PublicKeyCredentialType/public-key}}",
+        or the [=authenticator=] did not support any of the specified cryptographic parameters in {{PublicKeyCredentialCreationOptions/pubKeyCredParams}}.
+
+    :   <dfn>SecurityError</dfn>
+    ::  The [=effective domain=] was not a [=valid domain=],
+        or {{PublicKeyCredentialCreationOptions/rp}}.{{PublicKeyCredentialRpEntity/id}} was not a registrable domain suffix of nor was equal to the [=effective domain=].
+
+    :   <dfn>TypeError</dfn>
+    ::  The value of {{PublicKeyCredentialCreationOptions/user}}.{{PublicKeyCredentialUserEntity/id}} was not between 1 and 64 bytes (inclusive.)
```

Comment:
`TypeError` is thrown in other type error cases as well, like passing a string in place of a byte array or an array in place of an object (but notably not passing a byte array in place of a string - I assume it gets implicitly `.toString()`ed):

```js
navigator.credentials.create({
  publicKey: {
    rp: { name: "Test" },
    user: { id: "1234", name: "Test", displayName: "Test" },
    pubKeyCredParams: [{ type: "public-key", alg: -7 }],
    challenge: crypto.getRandomValues(new Uint8Array(32)),
  }
})

Uncaught (in promise) TypeError: Failed to execute 'create' on 'CredentialsContainer': Failed to read the 'publicKey' property from 'CredentialCreationOptions': Failed to read the 'user' property from 'PublicKeyCredentialCreationOptions': Failed to read the 'id' property from 'PublicKeyCredentialUserEntity': The provided value is not of type '(ArrayBuffer or ArrayBufferView)'.

```

Reply:
Yeah, I agree we shouldn't reference implementation-specific behaviours in the spec (unless we're specifying that behaviour as canonical and required), but this is rather an implicit consequence of converting JS values to WebIDL values, which is a specified (if convoluted) process. If I'm reading WebIDL right, I think this particular `TypeError` comes from the JS bindings for [Operations](https://webidl.spec.whatwg.org/#js-operations), which reference [creating an operation function](https://webidl.spec.whatwg.org/#dfn-create-operation-function), where step 2.1.5 invokes the [overload resolution algorithm](https://webidl.spec.whatwg.org/#dfn-overload-resolution-algorithm) whose step 12.19 throws this `TypeError` after failing to convert a JS argument value into a WebIDL value.

Anyway, to be clear, I certainly don't think we should enumerate every particular way a type mismatch could cause a `TypeError`, but I think it's worth mentioning that `TypeError` may be thrown not only specifically because of a bad `user.id` length. Maybe something like this?

```
    :   {{TypeError}}
    ::  The argument was not a valid {{CredentialCreationOptions}} value,
        or the value of <code>{{PublicKeyCredentialCreationOptions/user}}.{{PublicKeyCredentialUserEntity/id}}</code> was empty or longer than 64 bytes.
```

Reply:
Not a bad idea, I'll go ahead and incorporate this.

Hmm, `{{CredentialCreationOptions}}` links to the credman version of the type, how can I link it to our extended version instead? 🤔

Reply:
I don't think you can link directly to the `partial dictionary` WebIDL, but I guess you could link to [§5.1.1. CredentialCreationOptions Dictionary Extension](https://w3c.github.io/webauthn/#sctn-credentialcreationoptions-extension) instead?

Reply:
Sweet, I figured out how to link to that anchor while still having it say "CredentialCreationOptions". This is exactly what I wanted, thanks for the tip.

![Screenshot 2024-06-21 at 10 05 11 AM](https://github.com/w3c/webauthn/assets/5166470/a45129e4-8007-4b89-a0bb-6bff82984fa4)

Reply:
@emlun This is ready for your review again ✌️